Src/HALALMock/Services/Communication/Ethernet/TCP/Socket.cpp

```diff
@@ -101,185 +177,104 @@ void Socket::reset(){
 		connecting_sockets[remote_node] = this;
 	}
 	state = INACTIVE;
-	tcp_abort(connection_control_block);
-	connection_control_block = tcp_new();
-
-	tcp_bind(connection_control_block, &local_ip.address, local_port);
-	tcp_nagle_disable(connection_control_block);
-	tcp_arg(connection_control_block, this);
-	tcp_poll(connection_control_block,connection_poll_callback,1);
-	tcp_err(connection_control_block, connection_error_callback);
-
-	tcp_connect(connection_control_block, &remote_ip.address , remote_port, connect_callback);
+	close();
+	tcp_connection_sim();
+	
 
 }
 
 
 void Socket::send(){
-	pbuf* temporal_packet_buffer;
-	err_t error = ERR_OK;
-	while(error == ERR_OK && !tx_packet_buffer.empty() && tx_packet_buffer.front()->len <= tcp_sndbuf(socket_control_block)){
-		temporal_packet_buffer = tx_packet_buffer.front();
-		error = tcp_write(socket_control_block, temporal_packet_buffer->payload, temporal_packet_buffer->len, TCP_WRITE_FLAG_COPY);
-		if(error == ERR_OK){
-			tx_packet_buffer.pop();
-			tcp_output(socket_control_block);
-			memp_free_pool(memp_pools[PBUF_POOL_MEMORY_DESC_POSITION],temporal_packet_buffer);
-		}else{
-			if(error == ERR_MEM){
-				close();
-				ErrorHandler("Too many unacked messages on client socket, disconnecting...");
-			}else{
-				ErrorHandler("Cannot write to client socket. Error code: %d",error);
-			}
-		}
-	}
+	while (!tx_packet_buffer.empty()) {
+        vector<uint8_t> packet = tx_packet_buffer.front();
+        ssize_t sent_bytes = ::send(socket_fd, packet.data(), packet.size(), 0);
+        if (sent_bytes < 0) {
+            std::cerr << "Error sending packet\n";
+            return;
+        }
+        tx_packet_buffer.pop();
+    }
 }
 
+void Socket::start_receiving(){
+	is_receiving = true;
+    receiving_thread = std::thread(&Socket::receive, this);
+}
+void Socket::stop_receiving() {
+    if (is_receiving) {
+        is_receiving = false;
+        if (receiving_thread.joinable()) {
+            receiving_thread.join(); // wait until the thread finish
+        }
+    }
+}
+void Socket::receive() {
+    while (is_receiving) {
+        uint8_t buffer[1024]; // Buffer for the data
+        ssize_t received_bytes = ::recv(socket_fd, buffer, sizeof(buffer), 0);
+        if (received_bytes > 0) {
+            std::vector<uint8_t> packet(buffer, buffer + received_bytes); 
+            {
+                std::lock_guard<std::mutex> lock(mtx); 
+                rx_packet_buffer.push(packet);
+				process_data(); 
+            }
```

Comment:
Fun fact, TCP works in streams and we are using nonblocking mode, so the previous call to received will return all the packets that have arrived and can even return half of a packet in one call and the other half in the next. I don't know how the library was designed, but if someone made the assumption that the data passed to the packet constructor or Order::process_data was just one full packet (like UDP does it) this won't work and it will be a pain in the ass to fix, just letting everyone know.

Reply:
not really related to this PR, but it's important to know

Reply:
the packets need to match EXACTLY, otherwise bad things will happen

Reply:
we might have to change something in the impl actually